### PR TITLE
Add RAPIDS + Optuna simple example

### PR DESCRIPTION
**Pull request by Nanthini10**
## Motivation
A simple example to illustrate how to use RAPIDS with Optuna for HPO tasks

## Description of the changes
Added a script to `examples/optuna_rapids.py`

### CODE REVIEW

File: examples/optuna_rapids.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    n_estimators = trial.suggest_int("n_estimators", 100, 1000)
+    score = train_and_eval(X_param, y_param, max_depth=max_depth,
+                           n_estimators=n_estimators)
+    return score
+
+
+if __name__ == "__main__":
+    N_TRIALS = 10
+    study_name = "optuna_experiment"
+    storage_name = "sqlite:///simple_storage.db"
```
Comment (crcrpar) [inline examples/optuna_rapids.py:46]: [question] Can we run this example without an RDB storage?

**Review by crcrpar — COMMENTED** [review]
Thank you for your work. It basically looks great but I have some comments

1. Could you change the name of this file? For consistency with the other examples, I would name this "rapids_simple.py".
2. Could you run style checkers, i.e., black and isort like `black .` and `isort .`?
3. Could you skip this example in CI since we don't have any CI environment with GPUs? What I'd like you to do is add this filename to `IGNORES` in https://github.com/optuna/optuna/blob/6ead862f2bf516435a735ad0a1052a374cd97f74/.github/workflows/examples.yml#L49-L53.

File: examples/optuna_rapids.py (hunk 1 above)
Comment (Nanthini10) [reply examples/optuna_rapids.py:46]: Yes, we can. Do you want me to change that?

File: examples/optuna_rapids.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    score = accuracy_score(y_valid, y_pred)
+    return score
+
+
+def objective(trial, X_param, y_param):
+    max_depth = trial.suggest_int("max_depth", 7, 15)
+    n_estimators = trial.suggest_int("n_estimators", 100, 1000)
+    score = train_and_eval(X_param, y_param, max_depth=max_depth,
+                           n_estimators=n_estimators)
+    return score
```
[the remark this answers is not in the document]
Comment (Nanthini10) [reply examples/optuna_rapids.py:40]: These are generally the recommended range for these parameters, and I went with the top 2 most useful parameters (max_depth and n_estimators)

File: examples/rapids_simple.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
+"""
+Optuna example using RAPIDS library for optimization.
+
+In this example, we perform HPO on Iris dataset using cuML's
+RandomForestClassifier. This should be used a starting point
+to extend the search for larger problems, and wider depths.
+
+To run this example:
+
+    $ python optuna_rapids.py
```
Comment (crcrpar) [inline examples/rapids_simple.py:10]:
```suggestion
    $ python rapids_simple.py
```

File: examples/rapids_simple.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
+"""
+Optuna example using RAPIDS library for optimization.
+
+In this example, we perform HPO on Iris dataset using cuML's
```
Comment (crcrpar) [inline examples/rapids_simple.py:4]:
```suggestion
In this example, we perform hyperparameter optimization on Iris dataset using cuML's
```

File: examples/rapids_simple.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+RandomForestClassifier. This should be used a starting point
```
Comment (crcrpar) [inline examples/rapids_simple.py:5]: I'm not quite sure but this should be rather...
```suggestion
RandomForestClassifier. This should be used as a starting point
```

**Review by crcrpar — COMMENTED** [review]
Thank you for the updates!

I left nit-picking-ish comments 🙂

**Review by crcrpar — APPROVED** [review]
I tried this example locally and it worked well from the below log 😃 

```
py37 ❯ python examples/rapids_simple.py
[I 2020-10-17 21:16:35,316] A new study created in memory with name: rapids_experiment
cudf/utilities/bit.hpp(19): warning: cassert: [jitify] File not found
cudf/fixed_point/fixed_point.hpp(27): warning: cassert: [jitify] File not found
[I 2020-10-17 21:16:43,436] Trial 0 finished with value: 0.9736841917037964 and parameters: {'max_depth': 8, 'n_estimators': 189}. Best is trial 0 with value: 0.9736841917037964.
[I 2020-10-17 21:16:46,603] Trial 1 finished with value: 0.8947368264198303 and parameters: {'max_depth': 7, 'n_estimators': 370}. Best is trial 0 with value: 0.9736841917037964.
[I 2020-10-17 21:16:48,221] Trial 2 finished with value: 0.8947368264198303 and parameters: {'max_depth': 12, 'n_estimators': 172}. Best is trial 0 with value: 0.9736841917037964.
[I 2020-10-17 21:16:54,663] Trial 3 finished with value: 0.8947368264198303 and parameters: {'max_depth': 12, 'n_estimators': 754}. Best is trial 0 with value: 0.9736841917037964.
[I 2020-10-17 21:17:00,497] Trial 4 finished with value: 0.8947368264198303 and parameters: {'max_depth': 11, 'n_estimators': 644}. Best is trial 0 with value: 0.9736841917037964.
[I 2020-10-17 21:17:07,410] Trial 5 finished with value: 0.8947368264198303 and parameters: {'max_depth': 14, 'n_estimators': 807}. Best is trial 0 with value: 0.9736841917037964.
[I 2020-10-17 21:17:11,132] Trial 6 finished with value: 0.9736841917037964 and parameters: {'max_depth': 15, 'n_estimators': 372}. Best is trial 0 with value: 0.9736841917037964.
[I 2020-10-17 21:17:19,256] Trial 7 finished with value: 0.9736841917037964 and parameters: {'max_depth': 9, 'n_estimators': 897}. Best is trial 0 with value: 0.9736841917037964.
[I 2020-10-17 21:17:28,120] Trial 8 finished with value: 0.8947368264198303 and parameters: {'max_depth': 8, 'n_estimators': 898}. Best is trial 0 with value: 0.9736841917037964.
[I 2020-10-17 21:17:31,822] Trial 9 finished with value: 0.9210526347160339 and parameters: {'max_depth': 8, 'n_estimators': 401}. Best is trial 0 with value: 0.9736841917037964.
Number of finished trials: 10
Best trial:
  Value: 0.9736841917037964
  Params:
    max_depth: 8
    n_estimators: 189
```

**Comment by Nanthini10** [comment]
> I tried this example locally and it worked well from the below log 😃
> 
> ```
> py37 ❯ python examples/rapids_simple.py
> [I 2020-10-17 21:16:35,316] A new study created in memory with name: rapids_experiment
> cudf/utilities/bit.hpp(19): warning: cassert: [jitify] File not found
> cudf/fixed_point/fixed_point.hpp(27): warning: cassert: [jitify] File not found
> [I 2020-10-17 21:16:43,436] Trial 0 finished with value: 0.9736841917037964 and parameters: {'max_depth': 8, 'n_estimators': 189}. Best is trial 0 with value: 0.9736841917037964.
> [I 2020-10-17 21:16:46,603] Trial 1 finished with value: 0.8947368264198303 and parameters: {'max_depth': 7, 'n_estimators': 370}. Best is trial 0 with value: 0.9736841917037964.
> [I 2020-10-17 21:16:48,221] Trial 2 finished with value: 0.8947368264198303 and parameters: {'max_depth': 12, 'n_estimators': 172}. Best is trial 0 with value: 0.9736841917037964.
> [I 2020-10-17 21:16:54,663] Trial 3 finished with value: 0.8947368264198303 and parameters: {'max_depth': 12, 'n_estimators': 754}. Best is trial 0 with value: 0.9736841917037964.
> [I 2020-10-17 21:17:00,497] Trial 4 finished with value: 0.8947368264198303 and parameters: {'max_depth': 11, 'n_estimators': 644}. Best is trial 0 with value: 0.9736841917037964.
> [I 2020-10-17 21:17:07,410] Trial 5 finished with value: 0.8947368264198303 and parameters: {'max_depth': 14, 'n_estimators': 807}. Best is trial 0 with value: 0.9736841917037964.
> [I 2020-10-17 21:17:11,132] Trial 6 finished with value: 0.9736841917037964 and parameters: {'max_depth': 15, 'n_estimators': 372}. Best is trial 0 with value: 0.9736841917037964.
> [I 2020-10-17 21:17:19,256] Trial 7 finished with value: 0.9736841917037964 and parameters: {'max_depth': 9, 'n_estimators': 897}. Best is trial 0 with value: 0.9736841917037964.
> [I 2020-10-17 21:17:28,120] Trial 8 finished with value: 0.8947368264198303 and parameters: {'max_depth': 8, 'n_estimators': 898}. Best is trial 0 with value: 0.9736841917037964.
> [I 2020-10-17 21:17:31,822] Trial 9 finished with value: 0.9210526347160339 and parameters: {'max_depth': 8, 'n_estimators': 401}. Best is trial 0 with value: 0.9736841917037964.
> Number of finished trials: 10
> Best trial:
>   Value: 0.9736841917037964
>   Params:
>     max_depth: 8
>     n_estimators: 189
> ```

Thank You!

**Review by smly — COMMENTED** [review]
Thank you for your PR! I have some minor suggestions:

1. Please add this example to the "Examples with ML Libraries" list in `examples/README.md`.
https://github.com/optuna/optuna/tree/master/examples
2. How about changing `N_TRIALS` from 10 to 100? By default, `TPESampler` does random sampling when `N_TRIALS` is small. `N_TRIALS = 100` is more practical and consistent as an example code.

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your great PR. I successfully executed the example using `rapids-0.16` and `cudatoolkit=11.0`.

```console
[I 2020-10-23 14:40:15,218] Trial 9 finished with value: 0.9473684430122375 and parameters: {'max_depth': 11, 'n_estimators': 825}. Best is trial 5 with value: 1.0.
Number of finished trials: 10
Best trial:
  Value: 1.0
  Params:
    max_depth: 9
    n_estimators: 688
```

I agree with the suggestions (https://github.com/optuna/optuna/pull/1924#pullrequestreview-511490787), but we can work on them in a follow-up PR. 
@Nanthini10 What do you think?

**Comment by smly** [comment]
I agree with @toshihikoyanase's suggestion. This PR basically looks good to me. Thanks again!

**Review by toshihikoyanase — APPROVED** [review]
@smly Thank you for your reply. Let me merge this PR.

@Nanthini10 Thank you for your contribution. I'll create a follow-up PR, and please feel free to add comments on it.

**Comment by Nanthini10** [comment]
@smly @toshihikoyanase  Sorry for the late reply, I was OOTO. I agree with the suggestions too, we can follow up with a PR later.

**Pull request merged**